src/Conf.ml

```diff
@@ -1865,22 +1882,31 @@ let update_using_env conf =
                 , Sexp.List [Sexp.Atom name; Sexp.Atom reason] )))
   with Sys_error _ -> conf
 
-type 'a input = {kind: 'a; name: string; file: string; conf: t}
+type 'a input = {kind: 'a; name: string; file: file; conf: t}
 
 type action =
   | In_out of [`Impl | `Intf | `Use_file] input * string option
   | Inplace of [`Impl | `Intf | `Use_file] input list
   | Check of [`Impl | `Intf | `Use_file] input list
 
-let kind_of fname =
+let kind_of file =
   match !kind with
   | Some kind -> kind
   | None -> (
-    match Filename.extension fname with
-    | ".ml" -> `Impl
-    | ".mli" -> `Intf
-    | ".mlt" -> `Use_file
-    | _ -> `Impl )
+    match file with
+    | Stdin -> impossible "checked by validate"
```

Comment:
I don't like having this unchecked. We could have the kind and display name of a file computed before validate and passed safely (eg. `Stdin of kind | File of name * kind`). This would be a big change on how all this code works, so we should do it in an other PR.

Reply:
I agree but this assumes the kind of file arg is passed before the input files, otherwise it needs to be done afterwards and it's not much cleaner than what is currently done.